text/0000-dyno.md

```diff
@@ -0,0 +1,294 @@
+
+In `provide_context`, an error type provides access to data via a `Requisition` object, e.g., `req.provide_ref::<Backtrace>(&self.backtrace)`. The type of the reference passed to `provide_ref` is important here (and we encourage users to use explicit types with turbofish syntax even though it is not necessary). When a user calls `get_context_ref`, the requested type must match the type of an argument to `provide_ref`, e.g., the type of `&self.backtrace` is `&Backtrace`, so a call to `get_context_ref::<Backtrace>()` will return a reference to `self.backtrace`. An implementer can make multiple calls to `provide_ref` to provide multiple data with different types.
+
+Note that `Requisition` has methods for providing values as well as references, and for providing more complex types. These will be covered in the next section.
+
+The important parts of `provide_any` are
+
+```rust
+pub mod provide_any {
+    pub trait Provider {
+        fn provide<'a>(&'a self, req: Requisition<'a, '_>);
+    }
+
+    pub fn request_by_type_tag<'a, I: TypeTag<'a>>(provider: &'a dyn Provider) -> Option<I::Type> { ... }
+
+    pub struct Requisition<'a, 'b>(...);
+
+    impl<'a, 'b> Requisition<'a, 'b> {
+        pub fn provide_ref<T: ?Sized + 'static>(&mut self, value: &'a T) -> &mut Self { ... }
+        ...
+    }
+
+    pub trait TypeTag<'a>: Sized + 'static {
+        type Type: 'a;
+    }
+
+    pub mod tags { ... }
+}
+```
+
+We have mostly covered how this interface is used above. The last piece is the concept of type tags, as used as a type parameter to `request_by_type_tag`.
+
+`core::any::TypeId` is a unique identifier for a type, but can only be used with types without lifetimes (or with only `'static` lifetimes). A type tag is an object which provides a layer of indirection to work around this limitation. For example, consider the type `&'l String`. It does not have a `TypeId` because of the `'l` lifetime. But we can use a type tag, `Ref<String>` to represent `&'l String` and `Ref<String>` does have a `TypeId`.
+
+The `TypeTag` trait is an abstraction over all type tags. It does not have any methods, only an associated type for the type which the tag represents. E.g., `<Ref<T> as TypeTag<'a>>::Type` is `&'a T`.
+
+There is no universal type tag. A concrete type tag must be written for a 'category' of types. A few common tags are provided in `provide_any::tags`, including `Value` for any type bounded by `'static`, and `Ref` for types of the form `&'a T` where `T: 'static`. For less common types, the user must provide a tag which implements `TypeTag`; in this way the `provide_any` API generalises to all types.
```

Comment:
Why is `Ref` not a type tag combinator, where if `T` has a type tag then `Ref` can be used as type tag for references to `T` (somewhat similar to the `Option` combinator that does seem to exist)?

Reply:
My guess is convenience. I think if `Ref` was implemented the same way `Optional` is you'd have to invoke it with `foo.get_by_tag::<Ref<Value<MyType>>>()`. Thinking about it this doesn't seem too bad, since we already have `get_context_ref` that handles this case, so we may never need to use the `Ref` type tag directly as is, which seems like a good justification for updating it to compose with other TypeTags.

Reply:
One reason is that type tags represent `Sized` types, so using the compositional form for `Ref` would mean we lose the ability to represent references to `?Sized` types. You can add the `?Sized` bound to the `Type` associated type, but that means a lot of `I::Type: Sized` bounds all over the place, so it seems sub-optimal.